src/aspire/source/__init__.py

```diff
@@ -113,6 +115,7 @@ def __init__(self, L, n, dtype="double", metadata=None, memory=None):
 
         self.unique_filters = None
         self.generation_pipeline = Pipeline(xforms=None, memory=memory)
+        self._metadata_updated = None
```

Comment:
I'm not sure about adding this extra field here. It becomes harder to keep track of state. It's better to do as we discussed and have `save_metadata` return a list of tuples `(filename, index)` for each image that can then be passed to `save_images`.

Reply:
The `self._metadata_updated` is used to save the updated information such as (`_rlnImageName` column) for output the star file. We talked out that direct update to  `self._metadata` might be better, not using the list of tuples `(filename, index)`.  Since `self._metadata` also include other columns in addition to Relion columns and we tend to keep ImageSource immutable, I think creating another one is better. I changed the name to  `self._metadata_out` for output STAR file. In this way we also apply to the change of other columns such as rotation angles.

Reply:
Ok. I agree that we should keep the object immutable. The solution in that case is to go back to the previous solution, which is to return the list of tuples and take that as an argument to `save_images`. Adding another field `_metadata_out` to the object mutates it. So let's return the list of tuples instead.

Reply:
Change to use filename list in commit 4dd3e5f.